04-cs-fundamentals/internship/binary-search-trees.md

```diff
@@ -1,23 +1,253 @@
+
+On the other hand this is an unbalanced tree.
+
+![unbalanced bst](images/unbalanced-bst.png)
+
+<details>
+  <summary>How many nodes do you need to examine to find 5, in the 1st example.  How many in the 2nd?</summary>
+
+  In the 1st example you look at 25, then 10, and then 5.  In the 2nd you look at 100, then 50, then 25, then 15, then 10 and then 5.  
+
+  So the 1st requires 3 lookups while the 2nd requires 6.
+</details>
+
+<br />
+
+<details>
+  <summary>Based on your expert knowledge of Big-O.  What's the time complexity to look up a node in a balanced tree?</summary>
+
+  O(log n)
+</details>
+
+<details>
+  <summary>What about an unbalanced tree?</summary>
+
+  O(n)
+</details>
+
+Notice if a tree is balanced, when you move left or right, you eliminate half of the possible nodes.  This means you are essentially doing **binary search.**  If the tree is unbalanced, it more resembles linear search.  In general you can assume a tree is balanced given no other information just because if elements are inserted randomly, the resultant tree is likely to be more or less balanced.
+
+**Self-Balancing Trees** There are a lot of algorithms for [keeping a tree balanced](https://en.wikipedia.org/wiki/Self-balancing_binary_search_tree).  These are wonderful things to learn, but beyond the scope of this class.
+
+## Traversals
+
+A _traversal_ is an action visiting each node in a graph such as a tree.  There are several kinds of traversals, Breadth First Traversals which visit each node level, by level and Depth First Traversals which visit a node's children before it's siblings.
+
+### Breadth first Traversals
+
+When you conduct a _breadth-first traversal_ you visit each node by level.  In other words you explore nodes on a given level before exploring their children.  
+
+![Binary Search Tree 2](images/bst2.png)
+
+A Breadth First Traversal on the above tree could look like:
+
+[50, 25, 75, 10, 30, 60, 100]
+
+**Exercise**
+
+Do a breadth first traversal on the tree you created for:  
+"Ringo", "John", "Paul", and "George"
+
+### Depth First Traversals
+
+In a _Depth-first traversal_ you explore the children and grandchildren of a node before moving to it's sibling.
+
+There are three standard types of depth-first traversals:
+
+- **Pre-Order**:  Root, Left, Right
+- **In-Order**: Left, Root, Right
+- **Post-Order**: Left, Right, Root
+
+![Binary Search Tree 2](images/bst2.png)
+
+For the above Binary Search Tree
+- **Pre-Order**:  [50, 25, 10, 30, 75, 60, 100]
+- **In-Order**: [10, 25, 30, 50, 60, 75, 100]
+- **Post-Order**: [10, 30, 25, 60, 100, 75, 50]
+
+**Exercise**
+
+![bst3](images/bst3.png)
+
+<details>
+  <summary>In what order are the nodes in this tree encountered for each type of traversal
+  </summary>
+    Pre-Order  [17, 14, 20, 19, 52]
+    In-Order: [14, 17, 19, 20, 52]
+    Post-Order: [10, 19, 52, 20, 17]
+</details>
+
+<details>
+  <summary>What is the height of the above BST?
+  </summary>
+    3
```

Comment:
I love these questions! I would probably ask about the traversal orders last though.